finat/tensor_product.py

```diff
@@ -167,9 +167,15 @@ def dual_basis(self):
         # Outer product the dual bases of the factors
         qs, pss = zip(*(factor.dual_basis for factor in self.factors))
         ps = TensorPointSet(pss)
-        qis = tuple(q[gem.indices(len(q.shape))] for q in qs)
-        indices = tuple(chain(*(q.index_ordering() for q in qis)))
-        Q = gem.ComponentTensor(reduce(gem.Product, qis), indices)
+        alphas = [factor.get_indices() for factor in self.factors]
+        zetas = [factor.get_value_indices() for factor in self.factors]
+        # Index the factors by so that we can reshape into index-shape
+        # followed by value-shape
+        qis = [q[alpha + zeta] for q, alpha, zeta in zip(qs, alphas, zetas)]
+        Q = gem.ComponentTensor(
+            reduce(gem.Product, qis),
+            tuple(chain(*(alphas + zetas)))
+        )
```

Comment:
Looks good to me, though the naming of `alphas` suggests it refers to a geometric shape when really its just the indices over all the bases. Perhaps rename? (Or tell me I've misunderstood!)

Reply:
I'm just mimicking the naming in `_merge_evaluations`.

Reply:
In which case please say as much or copy across the explanatory comments from `_merge_evaluations`

Reply:
Done.